### PR TITLE
[saasherder] GQL classes and type hints

### PR DESCRIPTION
**Pull request by chassing**
2nd try after reverting https://github.com/app-sre/qontract-reconcile/pull/3296

This PR fixes the `AutoPromoter` class. The `promotions` attribute must be json serializable because it's sent to SQS.

### CODE REVIEW

File: reconcile/utils/mr/auto_promoter.py [hunk 1]
```diff
@@ -32,10 +36,20 @@ class ParentSaasConfigPromotion:
...
-    def __init__(self, promotions):
-        self.promotions = [p for p in promotions if p]
+    def __init__(
+        self, promotions: Union[Sequence[Promotion], Sequence[dict[str, Any]]]
+    ):
+        # !!! Attention !!!
+        # AutoPromoter is also initialized with promitions as dict by 'gitlab_mr_sqs_consumer'
+        # loaded from SQS message body, therefore self.promotions must be json serializable
+        self.promotions = [
+            p.dict() if isinstance(p, Promotion) else p for p in promotions
```
Comment (fishi0x01) [inline reconcile/utils/mr/auto_promoter.py:46]: Not sure, but would it be safer here to do `dict(by_alias=True)`?

**Review by fishi0x01 — APPROVED** [review]

File: reconcile/utils/mr/auto_promoter.py (hunk 1 above)
Comment (chassing) [reply reconcile/utils/mr/auto_promoter.py:46]: `Promotion` isn't a qenerate generated GQL class, and it doesn't use aliases.

Comment (fishi0x01) [reply reconcile/utils/mr/auto_promoter.py:46]: Was wondering because it has some Protocol class as one attribute. Can't that potentially be a problem?

Comment (chassing) [reply reconcile/utils/mr/auto_promoter.py]: Good point! Thanks for that! 🙇 The `json.dump` was working, but not the `json.load` :(
I reworked this stuff and introduced a `SaasPromotion` Protocol, set `by_alias`, and extended the test fixtures.

**Pull request merged**